R/epichains.R

```diff
@@ -0,0 +1,520 @@
+#'
+#' @param x An R object
+#'
+#' @return logical, `TRUE` if the object is an `<epichains_tree>` and `FALSE`
+#' otherwise
+#' @author James M. Azam
+#' @export
+is_epichains_tree <- function(x) {
+  inherits(x, "epichains_tree")
+}
+
+#' Test if x is an `epichains_summary` object
+#'
+#' @param x An R object
+#'
+#' @return logical, `TRUE` if the object is an `epichains_summary` and `FALSE`
+#' otherwise
+#' @author James M. Azam
+#' @export
+is_epichains_summary <- function(x) {
+  inherits(x, "epichains_summary")
+}
+
+#' Validate an `<epichains_tree>` object
+#'
+#' @param x An `<epichains_tree>` object
+#'
+#' @return No return.
+#' @author James M. Azam
+#' @export
+validate_epichains_tree <- function(x) {
+  if (!is_epichains_tree(x)) {
+    stop("Object must have an `<epichains_tree>` class")
+  }
+
+  # check for class invariants
+  stopifnot(
+    "object does not contain the correct columns" =
+      c("sim_id", "infector_id", "generation") %in%
+      colnames(x),
+    "column `sim_id` must be a numeric" =
+      is.numeric(x$sim_id),
+    "column `infector_id` must be a numeric" =
+      is.numeric(x$infector_id),
+    "column `generation` must be a numeric" =
+      is.numeric(x$generation)
+  )
+
+  invisible(x)
+}
+
+#' Validate an `<epichains_summary>` object
+#'
+#' @param x An `<epichains_summary>` object
+#'
+#' @return No return.
+#' @author James M. Azam
+#' @export
+validate_epichains_summary <- function(x) {
+  if (!is_epichains_summary(x)) {
+    stop("Object must have an `<epichains_summary>` class")
+  }
+
+  invisible(x)
+}
+
+#' `head` and `tail` method for `<epichains_tree>` class
+#'
+#' @param x An `<epichains_tree>` object
+#' @param ... further arguments passed to or from other methods
+#' @importFrom utils head
+#' @importFrom utils tail
+#' @return Object of class `data.frame`
+#' @author James M. Azam
+#' @export
+#' @details
+#' This returns the top rows of an `<epichains_tree>` object. Note that
+#' the object is originally sorted by `sim_id` and `infector_id` and the first
+#' unknown infectors (NA) have been dropped from
+#' printing method.
+#'
+#' To view the full output, use `as.data.frame(<object_name>)`.
+head.epichains_tree <- function(x, ...) {
+  # print head of the simulation output from the first known infector_id
+  x <- x[!is.na(x$infector_id), ]
+  utils::head(as.data.frame(x), ...)
+}
+
+#' @rdname head.epichains_tree
+#' @export
+tail.epichains_tree <- function(x, ...) {
+  utils::tail(as.data.frame(x), ...)
+}
+
+#' Aggregate cases in `<epichains_tree>` objects by "generation" or "time", if
+#' present
+#'
+#' @description
+#' This function provides a quick way to create a time series of cases over
+#' generation or time (if serials_dist was specified) from simulated
+#' `<epichains_tree>` objects.
+#'
+#' @param x An `<epichains_tree>` object.
+#' @param by The variable to aggregate by. Options include
+#' "time" and "generation".
+#' @param ... Other arguments passed to aggregate.
```

Comment:
```suggestion
#' @param ... ignored.
```

Although there may be value to actually pass it to `aggregate.data.frame()`?